libs/ledgerjs/packages/hw-transport-react-native-ble/ios/HwTransportReactNativeBle.swift

```diff
@@ -0,0 +1,217 @@
+//
+//  HwTransportReactNativeBle2.swift
+//  hw-transport-react-native-ble
+//
+//  Created by Juan CORTES on 23/6/22.
+//
+
+import Foundation
+import BleTransport
+import Bluejay
+
+enum TransportError: String, CaseIterable {
+    case bluetoothRequired = "bluetooth-required"
+    case deviceAlreadyConnected = "device-already-connected"
+    case deviceDisconnected = "device-disconnected"
+    case networkDown = "network-down"
+    case cantOpenDevice = "cant-open-device"
+    case pairingFailed = "pairingFailed"
+    case userPendingAction = "userPendingAction"
+    case writeError = "writeError"
+    case unknownError = "unknownError"
+}
+
+@objc(HwTransportReactNativeBle)
+class HwTransportReactNativeBle: RCTEventEmitter {
+    var rejectCallback: ((_ code: String, _ message: String, _ error: NSError?) -> Void)?
+    var queueTask: Queue?
+    
+    override init() {
+        super.init()
+        print(BleTransport.shared)
+        EventEmitter.sharedInstance.registerEventEmitter(eventEmitter: self)
+    }
+    
+    @objc func observeBluetooth() -> Void {
+        BleTransport.shared.bluetoothAvailabilityCallback { available in
+            EventEmitter.sharedInstance.dispatch(
+                Payload(
+                    event: Event.status.rawValue,
+                    type: available ? "PoweredOn" : "PoweredOff",
+                    data: nil
+                )
+            )
+        }
+    }
+    
+    @objc func listen(
+        _ resolve: @escaping RCTPromiseResolveBlock,
+        reject: RCTPromiseRejectBlock
+    ) -> Void {
+        if !BleTransport.shared.isBluetoothAvailable {
+            reject(TransportError.bluetoothRequired.rawValue, "", nil)
+            return
+        }
+        
+        DispatchQueue.main.async {
+            
+            BleTransport.shared.scan { discoveries in
+                let devices = discoveries.map{
+                    ExtraData(
+                        id: $0.peripheral.uuid.uuidString,
+                        rssi: $0.rssi,
+                        name: $0.peripheral.name,
+                        serviceUUIDs: [$0.serviceUUID.uuidString]
+                    )
+                }
+                
+                EventEmitter.sharedInstance.dispatch(
+                    Payload(
+                        event: Event.newDevices.rawValue,
+                        type: Event.newDevices.rawValue,
+                        data: ExtraData(
+                            devices: devices
+                        )
+                    )
+                )
+            } stopped: {_ in }
+        }
+    }
+    
+    @objc func stop(
+        _ resolve: @escaping RCTPromiseResolveBlock,
+        reject: RCTPromiseRejectBlock
+    ) -> Void {
+        if BleTransport.shared.isBluetoothAvailable {
+            DispatchQueue.main.async {
```

Comment:
BlueJay (the library internally used by the Native transport) requires the main thread, this will be droppable once @DantePuglisi moves us to the BluetoothCore or whatever is called implementation.